### PR TITLE
PlatformEngines: replace ad hoc stuff with Downloads & Tar stdlibs

### LINKED ISSUE #1830: Use libCURL instead of platform engines
_This is very speculative, so mainly for brainstorming_

Currently HTTTP downloads within Pkg are done using the `PlatformEngines` subsystem, which shell out to locally installed commandline tools to perform the download. This generally means the `curl`/`wget`/`fetch` executables on unix and `PowerShell` on windows. This code is reasonably complex and difficult to maintain (I think)

In order to make things streamlined, one thought is to replace this with HTTP.jl. However, that is a large library, and splitting it will be some effort. 

Another idea is to use libCurl. 

* libCURL is *widely* used and *very* stable. Every car, phone and TV has multiple instances. 
* It's very efficient, in CPU and memory -- it can do multiple downloads simultaneously in async mode. 
* There is now a well maintained libCURL_jll, and the recent jll work on base julia makes it easy to pull in. One important feature here is that the jll correctly builds against the mbedTLS jll
* libCURL has http2 support (and ftp support and a dozen other protocols ...)

A few concerns, not insurmountable:

* libCURL has a low level C API. A higher level Julia API [needs to be written](https://curl.haxx.se/libcurl/c/libcurl-tutorial.html), ideally using the [multi-api](https://curl.haxx.se/libcurl/c/libcurl-multi.html). There are julia examples using the [easy api](https://github.com/JuliaWeb/LibCURL.jl#example-fetch-a-url)
* proxies (potentially authenticated)  need to be handled 
* root certificates needs to be handled

cc: @StefanKarpinski @staticfloat

### PR DESCRIPTION
**Pull request by StefanKarpinski**
Fixes #1830

### CODE REVIEW

File: src/PlatformEngines.jl [hunk 1]
```diff
@@ -802,18 +682,7 @@ function download(
...
-    end
-    try
-        run(download_cmd, (devnull, verbose ? stdout : devnull, verbose ? stderr : devnull))
-    catch e
-        if isa(e, InterruptException)
-            rethrow()
-        end
-        error("Could not download $(url) to $(dest):\n$(e)")
-    end
+    Downloads.download(url, dest, headers=headers)
```
Comment (KristofferC) [inline src/PlatformEngines.jl:685]:
```suggestion
    Downloads.download(url, dest, headers)
```

**Comment by KristofferC** [comment]
I tried this with the tweak I commented and then it starts to download but I think some concurrent download tasks get stuck. I did an `instantiate` in my global env and it got really slow and then it just hung completely.

**Comment by StefanKarpinski** [comment]
Yeah, there's some kind of task deadlock issue with a lot of downloads. Definitely needs debugging.

**Comment by StefanKarpinski** [comment]
Inserting a `yield` before calling `Downloads.download` "fixes" the issue.

**Comment by KristofferC** [comment]
> Inserting a yield before calling Downloads.download "fixes" the issue.

I tried this on the latest nightly and it still hangs for me. It might be related but the throughput seems quite a lot slower as well (but maybe this is because tasks are hanging):

PR:
Screenshot-20201008170442-1119x132

1.5:

Screenshot-20201008170631-532x114

**Comment by StefanKarpinski** [comment]
So this mostly works but the throughput regression needs to be investigated and may be the cause of the timeouts on various CI runs (I restarted a couple of them).

**Comment by KristofferC** [comment]
You don't get the hang when running instantiate on a big environment? I'll try again later.

**Comment by KristofferC** [comment]
I made a benchmark script:

```
using Pkg
using Pkg.TOML
import Downloads
function gather_urls(env="v1.6")
    d = TOML.parsefile(joinpath(homedir(), ".julia/environments/$env/Manifest.toml"))
    regpath = joinpath(homedir(), ".julia/registries/General")
    reg = TOML.parsefile(joinpath(regpath, "Registry.toml"))
    urls = String[]
    for (pkg, infos) in d
        for info in infos
            uuid = info["uuid"]
            tree = get(info, "git-tree-sha1", nothing)
            tree === nothing && continue
            pkg = get(reg["packages"], uuid, nothing)
            pkg === nothing && continue
            path = pkg["path"]
            repo = TOML.parsefile(joinpath(regpath, path, "Package.toml"))["repo"]
            push!(urls, string(Pkg.pkg_server(), "/package/$uuid/$tree"))
        end
    end
    return urls
end
urls = gather_urls()
function download_urls(urls; num_concurrent=8, path=mktempdir(), old_way=true)
    Pkg.PlatformEngines.probe_platform_engines!()
    broke = false
    @sync for i in 1:num_concurrent
        @async while true
            broke && break
            isempty(urls) && break
            url = popfirst!(urls)
            @info "downloading $url"
            file = joinpath(path, basename(url))
            try
                if old_way
                    Pkg.PlatformEngines.download(url, file; verbose=false)
                else
                    Downloads.download(url, file)
                end
            catch e
                if e isa InterruptException
                    broke = true
                end
                @show e
                @warn "url failed to download"
            end
        end
    end
end
@time download_urls(urls; old_way=false)
```

and running this on my mac I get faster speed with `old_way=false` than `old_way=true`, that is faster with `Downloads`. So I am not sure why I got the bad throughput in Pkg, I'll have to test this on my linux machine tomorrow where I made the post above.

**Comment by KristofferC** [comment]
I also get the hang in Pkg on my mac, which runs the benchmark script fine. So it doesn't seem to be related to only running the downloads..

**Comment by KristofferC** [comment]
Tried it again with latest Downloads.jl and it downloads fast(!) but unfortunately there still seem to be some tasks getting stuck.

However, on the benchmark above it works fine and it is more than twice as fast as the old way.

**Comment by StefanKarpinski** [comment]
Awesome to hear that. I had tried it hoping the the hangs to go away, but alas, not yet. So the new speed is great, but I still need to investigate the hanging. I've already got a branch of downloads that prints a bunch of instrumented data in CSV format so I can see which downloads are getting stuck. Will continue with that tomorrow.

File: Project.toml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 name = "Pkg"
-uuid = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+uuid = "b4ba1df2-8887-4706-8585-f0048dd7ac84"
```
Comment (KristofferC) [inline Project.toml:2]:
```suggestion
uuid = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
```

**Comment by KristofferC** [comment]
bild

:tada:

**Pull request merged**